### PR TITLE
audacity: 2.4.2 -> 3.0.2

### PR DESCRIPTION
**Pull request by wlib**
The XGD-config patch is removed since it conflicts with the new version of audacity, as it was not merged upstream.

Audacity now requires their own fork of wxWidgets, according to https://wiki.audacityteam.org/wiki/Building_On_Linux - this is the cause of current trouble since I am new to nix and unsure how to integrate it. I copied the current 3.1.4 nixpkgs version of wxWidgets found at pkgs/development/libraries/wxwidgets/3.1/default.nix and modified it to retrieve the forked version, but that's it so far.

###### Motivation for this change
Audacity has not been updated to the latest major version (3) on any linux distribution that I found. That includes Ubuntu (of course), but also even Arch and Gentoo. This update is filled with bug fixes and should set us up to more easily update in the future as well.

###### Things done

The essence of the changes so far are noted in the commit message - this will not build. I am setting this up as a draft in order to invite help/suggestions to pull in wxWidgets correctly as a build dependency. I believe that this will be the only change needed to build this update correctly.

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cc: @lheckemann @peterhoeg

note: I closed the pull request that was from my master branch; this is a reopening with its own branch following what I have observed to be convention. The previous version was here: https://github.com/NixOS/nixpkgs/pull/121618 and it had two relevant comments that I left.

### CODE REVIEW

**Comment by wlib** [comment]
I'm not sure if this might be relevant, but https://github.com/audacity/audacity/blob/master/cmake-proxies/wxWidgets/CMakeLists.txt#L76 might be helpful to look at.

**Comment by peterhoeg** [comment]
For wxgtk, can't we just override it with the patches from audacity instead of copying the whole definition?

**Comment by wlib** [comment]
> For wxgtk, can't we just override it with the patches from audacity instead of copying the whole definition?

That is a good idea, I'm just not sure how to apply it. Does this look good?
```nix
# https://github.com/wxWidgets/wxWidgets/compare/v3.1.3...audacity:audacity-fixes-3.1.3
patches = [
  (fetchpatch {
    url = "https://github.com/wxWidgets/wxWidgets/compare/v3.1.3...audacity:07e7d832c7a337aedba3537b90b2c98c4d8e2985.patch";
    sha256 = "047xl6hwfpwr6j7yxrkwh1akf7cjch3lnk4jnwq29sgdq8s5s2wi";
    name = "audacity_wxWidgets_fixes.patch";
  })
];
```

File: pkgs/applications/audio/audacity/default.nix [hunk 1]
```diff
@@ -53,6 +54,19 @@ stdenv.mkDerivation rec {
     sha256 = "035qq2ff16cdl2cb9iply2bfjmhfl1dpscg79x6c9l0i9m8k41zj";
   };
 
+  patches = [
+    (fetchpatch {
+      url = "https://github.com/audacity/audacity/commit/007852e51fcbb5f1f359d112f28b8984a604dac6.patch";
```
Comment (wlib) [inline pkgs/applications/audio/audacity/default.nix:59]: Should this be audacity/audacity as it is now, or serebit/audacity?

**Comment by wlib** [comment]
It says that this works, but it won't build on my computer because I'm getting patch failures for some reason for https://github.com/wxWidgets/wxWidgets/compare/v3.1.3...audacity:07e7d832c7a337aedba3537b90b2c98c4d8e2985.patch

**Comment by peterhoeg** [comment]
An alternative is to override `src` to point to the audacity "fork" instead of adding the patches. Less error prone.

**Comment by wlib** [comment]
I didn't even consider that lol, trying it now

**Comment by serebit** [comment]
Nice to see my work being put to good use.

**Comment by wlib** [comment]
I applied this:
```
47 | let
48 |   wxGTK-audacity = wxGTK.overrideAttrs (oldAttrs: rec {
49 | -     patches = (oldAttrs.patches or []) ++ [
50 | -       (fetchpatch {
51 | -         url = "https://github.com/wxWidgets/wxWidgets/compare/v3.1.3...audacity:07e7d832c7a337aedba3537b90b2c98c4d8e2985.patch";
52 | -         sha256 = "00sgc7amgmymlgx8l3m9ls0cghmnbcfwbnslikridw6vlsp0bg08";
53 | -         name = "audacity_wxWidgets_fixes.patch";
54 | -       })
55 | -     ];
49 | +     src = fetchFromGitHub {
50 | +       owner = "audacity";
51 | +       repo = "wxWidgets";
52 | +       rev = "07e7d832c7a337aedba3537b90b2c98c4d8e2985";
53 | +       sha256 = "1mawnkcrmqj98jp0jxlnh9xkc950ca033ccb51c7035pzmi9if9a";
54 | +       fetchSubmodules = true;
55 | +     };
56 |   });
57 | in
58 | stdenv.mkDerivation rec {
```

and got this error in the middle of compilation:
```
/build/source/src/FileNames.cpp:815:10: fatal error: /usr/include/linux/magic.h: No such file or directory
  815 | #include "/usr/include/linux/magic.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by OPNA2608** [comment]
Add `linuxHeaders` to `nativeBuildInputs` and `substituteInPlace` that path in src/FileNames.cpp to just "linux/magic.h". Or you could substitute it with `${linuxHeaders}/include/linux/magic.h` instead but it really shouldn't be this hardcoded in the first place.

**Comment by wlib** [comment]
I have a successful build on my nixos desktop now

**Comment by Ma27** [comment]
LGTM :+1:

**Comment by wlib** [comment]
Thank you everyone

**Pull request merged**